README.txt
.. contents::

Introduction
============

Requirements
============

* docsplit and dependencies (http://documentcloud.github.com/docsplit/)
* install collective.documentvier in your Plone Site



CHANGES.txt
Changelog
=========

1.4 (unreleased)
----------------

- Nothing changed yet.


1.3 (2015-07-31)
----------------

- Don't install collective.documentviewer. We use the code but we
  don't want it installed because it overrides views for PDF files.
  [deroiste]


1.2 (2015-05-15)
----------------

- Nothing changed yet.


1.1 (2015-05-15)
----------------

- Nothing changed yet.


1.0a2 (2014-07-11)
------------------

- improved buildout [reinhardt]
- added some logging [reinhardt]
- added view to upload documents for conversion, refs #10398 [reinhardt]
- limit number of images to 20 by default, refs #10398 [reinhardt]
- new version of documentviewer [reinhardt]


1.0a1 (2012-08-31)
------------------

- initial release
  [reinhardt]


src/slc/docconv/tests/test_convert.py
import os
import unittest2 as unittest
from ZPublisher.HTTPRequest import FileUpload
from cgi import FieldStorage
from io import BytesIO
from plone import api
from slc.docconv.convert import _collect_data
from slc.docconv.testing import SLC_DOCCONV_INTEGRATION_TESTING
from zipfile import ZipFile


class TestUtils(unittest.TestCase):

    def test_collect_data(self):
        storage_dir = os.path.join(os.path.dirname(__file__), 'storage_dir')
        converted = _collect_data(storage_dir)

        self.assertIn('pdfs', converted)
        self.assertGreaterEqual(len(converted['pdfs']), 1)

        self.assertIn('thumbnails', converted)
        self.assertGreaterEqual(len(converted['thumbnails']), 1)

        self.assertIn('previews', converted)
        self.assertGreaterEqual(len(converted['previews']), 1)

        msg = 'incorrect sorting'
        thumb2 = open(os.path.join(storage_dir, 'small/dump_2.gif'), 'r')
        self.assertEqual(converted['thumbnails'][1], thumb2.read(), msg=msg)
        thumb2.close()
        prev2 = open(os.path.join(storage_dir, 'large/dump_2.gif'), 'r')
        self.assertEqual(converted['previews'][1], prev2.read(), msg=msg)
        prev2.close()


class TestConvert(unittest.TestCase):

    layer = SLC_DOCCONV_INTEGRATION_TESTING

    def setUp(self):
        self.portal = self.layer['portal']
        self.request = self.layer['request']

    def test_convert_empty(self):
        self.request.form = {}
        convert = api.content.get_view(
            context=self.portal,
            request=self.request,
            name='convert-external')
        convert()
        self.assertEqual(self.request.response.getStatus(), 415)

    def test_convert_file(self):
        in_file = open(os.path.join(os.path.dirname(__file__), 'test.odt'), 'r')
        env = {'REQUEST_METHOD': 'PUT'}
        headers = {'content-type': 'text/html',
                   'content-length': len(in_file.read()),
                   'content-disposition': 'attachment; filename=%s' % 'test.odt'}
        in_file.seek(0)
        fs = FieldStorage(fp=in_file, environ=env, headers=headers)
        f = FileUpload(fs)
        self.request.form = {
            'filedata': f,
        }
        convert = api.content.get_view(
            context=self.portal,
            request=self.request,
            name='convert-external')
        output = convert()
        stream = BytesIO(output)
        fzip = ZipFile(stream)

        pdfs = [x.filename for x in fzip.filelist
                if x.filename.endswith('.pdf')]
        self.assertGreaterEqual(len(pdfs), 1)
        self.assertGreaterEqual(len(fzip.read(pdfs[0])), 1)

        thumbs = [x.filename for x in fzip.filelist
                  if x.filename.startswith('small/')
                  and x.filename != 'small/']
        self.assertGreaterEqual(len(thumbs), 1)
        self.assertGreaterEqual(len(fzip.read(thumbs[0])), 1)

        previews = [x.filename for x in fzip.filelist
                    if x.filename.startswith('large/')
                    and x.filename != 'large/']
        self.assertGreaterEqual(len(previews), 1)
        self.assertGreaterEqual(len(fzip.read(previews[0])), 1)

        in_file.close()


src/slc/docconv/tests/test_example.py
import unittest2 as unittest


from slc.docconv.testing import\
    SLC_DOCCONV_INTEGRATION_TESTING


class TestExample(unittest.TestCase):

    layer = SLC_DOCCONV_INTEGRATION_TESTING
    
    def setUp(self):
        # you'll want to use this to set up anything you need for your tests 
        # below
        pass

    def test_success(self):
        sum = 1 + 3
        self.assertEquals(sum, 4)

    def test_failure(self):
        sum = 2 + 3
        self.assertEquals(sum, 4)


src/slc/docconv/testing.py
from plone.app.testing import PLONE_FIXTURE
from plone.app.testing import PloneSandboxLayer
from plone.app.testing import IntegrationTesting
from plone.app.testing import FunctionalTesting
from plone.app.testing import applyProfile

from zope.configuration import xmlconfig

class SlcDocconv(PloneSandboxLayer):

    defaultBases = (PLONE_FIXTURE, )

    def setUpZope(self, app, configurationContext):
        # Load ZCML for this package
        import slc.docconv
        xmlconfig.file('configure.zcml',
                       slc.docconv,
                       context=configurationContext)


    def setUpPloneSite(self, portal):
        pass

SLC_DOCCONV_FIXTURE = SlcDocconv()
SLC_DOCCONV_INTEGRATION_TESTING = \
    IntegrationTesting(bases=(SLC_DOCCONV_FIXTURE, ),
                       name="SlcDocconv:Integration")

src/slc/docconv/patches/patch_run_command.py
# Patch _run_command to allow setting of RLIMITS
import subprocess
from logging import getLogger
import resource
from collective.documentviewer import convert

logger = getLogger('collective.documentviewer')

MAX_CPU = 1800


def setlimits():
    """ Set the max CPU time allowed for a single converter process """
    rsrc = resource.RLIMIT_CPU
    # need to know current hard limit as we can't increase that
    soft, hard = resource.getrlimit(rsrc)
    if MAX_CPU > hard:
        soft = hard
    else:
        soft = MAX_CPU
    resource.setrlimit(rsrc, (soft, hard))
    logger.info("RLIMIT_CPU set to %s/%s" % (soft, hard))


def _run_command(self, cmd):
    """ copied over from documentviewer to be able to set resource limits
    """
    if isinstance(cmd, basestring):
        cmd = cmd.split()
    cmdformatted = ' '.join(cmd)
    logger.info("Running rlimited command %s" % cmdformatted)
    process = subprocess.Popen(cmd, stdout=subprocess.PIPE,
                               stderr=subprocess.PIPE,
                               close_fds=self.close_fds,
                               preexec_fn=setlimits)
    output, error = process.communicate()
    process.stdout.close()
    process.stderr.close()
    if process.returncode != 0:
        error = """Command
%s
finished with return code
%i
and output:
%s
%s""" % (cmdformatted, process.returncode, output, error)
        logger.info(error)
        raise Exception(error)
    logger.info("Finished Running Command %s" % cmdformatted)
    return output

logger.info("Patching _run_command to support RLIMIT")
convert.setlimits = setlimits
convert.BaseSubProcess._run_command = _run_command


src/slc/docconv/convert.py
import os
import resource
import shutil
import subprocess
from bs4 import BeautifulSoup
from five import grok
from io import BytesIO
from os import path, walk, remove
from zipfile import ZipFile
from Products.CMFPlone.interfaces import IPloneSiteRoot
from collective.documentviewer.settings import GlobalSettings
from collective.documentviewer.convert import DUMP_FILENAME
from collective.documentviewer.convert import DocSplitSubProcess
from collective.documentviewer.utils import mkdir_p
from logging import getLogger


logger = log = getLogger(__name__)

grok.templatedir('templates')

MAX_CPU = 1800


def setlimits():
    """ Set the max CPU time allowed for a single converter process """
    rsrc = resource.RLIMIT_CPU
    # need to know current hard limit as we can't increase that
    soft, hard = resource.getrlimit(rsrc)
    if MAX_CPU > hard:
        soft = hard
    else:
        soft = MAX_CPU
    resource.setrlimit(rsrc, (soft, hard))


class DocconvDocSplitSubProcess(DocSplitSubProcess):
    """Customised to limit the number of pages"""

    def dump_images(self, filepath, output_dir, sizes, format, lang='eng', limit=20):
        # docsplit images pdf.pdf --size 700x,300x,50x
        # --format gif --output
        pages = self.get_num_pages(filepath)
        if pages < limit:
            limit = pages
        cmd = [self.binary, "images", filepath,
            '--size', ','.join([str(s[1]) + 'x' for s in sizes]),
            '--format', format,
            '--rolling',
            '--output', output_dir,
            '--pages', '1-%s' % limit]

        self._run_command(cmd)

        # now, move images to correctly named folders
        for name, size in sizes:
            dest = os.path.join(output_dir, name)
            if os.path.exists(dest):
                shutil.rmtree(dest)

            source = os.path.join(output_dir, '%ix' % size)
            shutil.move(source, dest)

    def _run_command(self, cmd):
        """ copied over from documentviewer to be able to set resource limits
        """
        if isinstance(cmd, basestring):
            cmd = cmd.split()
        cmdformatted = ' '.join(cmd)
        logger.info("Running command %s" % cmdformatted)
        process = subprocess.Popen(cmd, stdout=subprocess.PIPE,
                                   stderr=subprocess.PIPE,
                                   close_fds=self.close_fds,
                                   preexec_fn=setlimits)
        output, error = process.communicate()
        process.stdout.close()
        process.stderr.close()
        if process.returncode != 0:
            error = """Command
%s
finished with return code
%i
and output:
%s
%s""" % (cmdformatted, process.returncode, output, error)
            logger.info(error)
            raise Exception(error)
        logger.info("Finished Running Command %s" % cmdformatted)
        return output


try:
    docsplit = DocconvDocSplitSubProcess()
except IOError:
    logger.exception("No docsplit installed. slc.docconv will not work.")
    docsplit = None


class ConvertExternal(grok.View):
    grok.name('convert-external')
    grok.context(IPloneSiteRoot)
    grok.require('zope2.View')

    def update(self):
        self.gsettings = GlobalSettings(self.context)

    def render(self):
        filedata = self.request.get('filedata')
        if not filedata:
            self.request.RESPONSE.setStatus(415)
            msg = 'No filedata found'
            log.warn(msg)
            return msg
        if not docsplit:
            self.request.RESPONSE.setStatus(500)
            msg = 'docsplit not found, check that docsplit is installed'
            log.error(msg)
            return msg

        fzipfilelist = []
        filename_base = filedata.filename.decode('utf8')
        if '.' in filename_base:
            filename_base = '.'.join(filename_base.split('.')[:-1])
        storage_dir = path.join(self.gsettings.storage_location, filename_base)
        if filedata.headers.get('content-type') == 'application/octetstream':
            filename_dump = path.join(self.gsettings.storage_location, '.'.join((filename_base, 'html')))
        else:
            filename_dump = path.join(self.gsettings.storage_location, filedata.filename.decode('utf8'))
            if filename_dump.endswith(filename_base):
                filename_dump = '.'.join([filename_dump, 'dat'])
        filename_pdf = path.join(storage_dir, 'converted.pdf') #'.'.join((filename_base, 'pdf')))

        if not path.exists(storage_dir):
            mkdir_p(storage_dir)
        if path.exists(filename_dump):
            remove(filename_dump)
        # do we have a zip file?
        if filedata.headers.get('content-type') == 'application/octetstream':
            # extract it
            stream = BytesIO(filedata.read())
            fzip = ZipFile(stream)
            fzipfilelist = fzip.filelist
            html = [x.filename for x in fzipfilelist if x.filename.endswith('.html') or x.filename.endswith('.htm')]
            if not html:
                self.request.RESPONSE.setStatus(415)
                msg = 'No html file found in zip'
                log.warn(msg)
                return msg
            fzip.extractall(self.gsettings.storage_location)
            shutil.move(path.join(self.gsettings.storage_location, html[0]), filename_dump)
            fzip.close()
            stream.close()

            # make img src paths absolute
            htmlfile = open(filename_dump, 'r')
            soup = BeautifulSoup(htmlfile.read())
            htmlfile.close()
            for img in soup.find_all('img'):
                if not img.has_key('src'):
                    continue
                img['src'] = path.join(self.gsettings.storage_location, img['src'])
            htmlfile = open(filename_dump, 'w')
            htmlfile.write(str(soup))
            htmlfile.close()

        else:
            fi = open(filename_dump, 'wb')
            fi.write(filedata.read())
            fi.close()

        if 'pdf' in filedata.headers.get('content-type'):
            shutil.move(filename_dump, filename_pdf)
        else:
            if path.exists(path.join(storage_dir, DUMP_FILENAME)):
                remove(path.join(storage_dir, DUMP_FILENAME))
            docsplit.convert_to_pdf(filename_dump, filename_dump, storage_dir)
            shutil.move(path.join(storage_dir, DUMP_FILENAME), filename_pdf)

        args = dict(sizes=(('large', self.gsettings.large_size),
                           ('normal', self.gsettings.normal_size),
                           ('small', self.gsettings.thumb_size)),
                ocr=self.gsettings.ocr,
                detect_text=self.gsettings.detect_text,
                format=self.gsettings.pdf_image_format,
                converttopdf=False,
                filename=filedata.filename.decode('utf8'),
                inputfilepath=filename_pdf)
        docsplit.convert(storage_dir, **args)

        stream = BytesIO()
        zipped = ZipFile(stream, 'w')
        for entry in walk(storage_dir):
            relpath = path.relpath(entry[0], storage_dir)
            if not entry[0] == storage_dir:
                # if it's not the top dir we want to add it
                zipped.write(entry[0], relpath.encode('CP437'))
            # we always want to add the contained files
            for filename in entry[2]:
                relative = path.join(relpath, filename)
                zipped.write(path.join(entry[0], filename), relative)
        zipped.close()
        zipdata = stream.getvalue()
        stream.close()
        shutil.rmtree(storage_dir)
        for ff in fzipfilelist:
            if not ff.filename == html[0]: # this one has already been consumed by convert_to_pdf
                remove(path.join(self.gsettings.storage_location, ff.filename))

        R = self.request.RESPONSE
        R.setHeader('content-type', 'application/zip')
        R.setHeader('content-disposition', 'inline; filename="%s"' % '.'.join((filename_base, u'zip')).encode('utf8'))
        R.setHeader('content-length', str(len(zipdata)))
        return zipdata


class ConvertUpload(grok.View):
    grok.name('convert-upload')
    grok.context(IPloneSiteRoot)
    grok.require('slc.docconv.convert')
    grok.template('convert')
